src/mlpack/methods/ann/gan.hpp

```diff
@@ -0,0 +1,181 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ * @author Shikhar Jaiswal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_HPP
+#define MLPACK_METHODS_ANN_GAN_HPP
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+#include <mlpack/methods/ann/visitor/reset_visitor.hpp>
+#include <mlpack/methods/ann/visitor/weight_size_visitor.hpp>
+#include <mlpack/methods/ann/visitor/weight_set_visitor.hpp>
+
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. **/ {
+template<typename Model, typename InitializationRuleType, class Noise>
+class GAN
+{
+ public:
+  /**
+   * Constructor for GAN class.
+   *
+   * @tparam Model The class type of Generator and Discriminator.
+   * @tparam InitializationRuleType Type of Initializer.
+   * @tparam Noise The noise function to use.
+   * @param trainData The real data.
+   * @param generator Generator network.
+   * @param discriminator Discriminator network.
+   * @param batchSize BatchSize to be used for training.
+   * @param generatorUpdateStep Number of steps of Discriminator training
+   *                            before updating generator.
+   * @param preTrainSize Num of preTraining step of Discriminator.
+   * @param multiplier Ratio of learning rate of Discriminator to the Generator.
+   */
+  GAN(arma::mat& trainData,
+      Model& generator,
+      Model& discriminator,
+      InitializationRuleType initializeRule,
+      Noise noiseFunction,
+      size_t noiseDim,
+      size_t batchSize,
+      size_t generatorUpdateStep,
+      size_t preTrainSize,
+      double multiplier);
+
+  // Reset function.
+  void Reset();
+
+  // Train function.
+  template<typename OptimizerType>
+  void Train(OptimizerType& Optimizer);
+
+  /**
+   * Evaluate function for the GAN gives the performance of the GAN on the
+   * current input.
+   *
+   * @param parameters The parameters of the network.
+   * @param i Index of the current input.
+   */
+  double Evaluate(const arma::mat& parameters, const size_t i,
```

Comment:
Picky style comment, the parameter should align with the first one. Same for the ``Gradient``method.